docs/design/2022-10-13-mysql-compatible-auto-increment.md

```diff
@@ -0,0 +1,110 @@
+
+The second one is, we can keep the old things unchanged, and introduce a new syntax. The benefit of this approach is that our users will never get into trouble when they upgrade TiDB.
+
+And another way, we can reuse the `AUTO_ID_CACHE 1` table option for it:
+
+```
+CREATE TABLE t (id int key AUTO_INCREMENT) AUTO_ID_CACHE 1;
+```
+
+TiDB extends the syntax to control the AUTO ID CACHE size use the `AUTO_ID_CACHE` table option. `AUTO_ID_CACHE 1` is equivalent to no caching. Now it's of the same semantic, just that the implementation is changed. I think the this approach is the best.
+
+## Implementation
+
+### Centralized auto ID allocating service
+
+This is the core part of the implementation. There will be a single-point process, responsible for allocating auto IDs.
+
+(To be exact, the allocating service will be embeded into the TiDB process currently for simplicity, maybe in the future it can be moved to a standalone process. But wherever it is, the idea is the same)
+
+For every `AUTO_INCREMENT` ID allocation, TiDB send a request to this process. The process will `+1` internally, so the ID is **unique**, **monotone increasing**, and **consecutive**.
+
+Grpc is used for the communication protocol. This part is relatively easy so I would not go to much details in this document.
+
+```
+message AutoIDRequest {
+    int64 dbID = 1;
+    int64 tblID = 2;
+    uint64 n = 3;
+    int64 increment = 4;
+    int64 offset = 5;
+}
+
+message AutoIDResponse {
+    int64 min = 1;
+    int64 max = 2;
+}
+
+service AutoIDAlloc {
+    rpc AllocAutoID(AutoIDRequest) returns (AutoIDResponse) {}
+}
+```
+
+### HA of the service
+
+To overcome the SPOF(single-point-of-failure) issue, the centralized auto ID allocating service should have at least a primary and a backup process. [Etcd](https://etcd.io/) is used for that.
+
+Both the primary and backup regist themself in the etcd, the elected primay process serves the workload. When it's gone (crash, quit or anything), the backup take up the role and provide the service. We can just use the pd embedded etcd, this is quite convenient.
+
+When the switch happen, the current max allocated ID information is required so as to guarantee the uniqueness. The new primary process allocate IDs begin from the max allocated ID plus one.
+
+We can persist the max allocated ID every time, but that's costly. Another choice is persisting it periodically, it's safe to allocate IDs in range `[base, max persisted)`. When the primary process crash abnormally, the backup process gets the max persisted ID as its base. Etcd is also used as the storage for the max persisted ID. This optimization could still make the ID not be consecutive, but it's not so common (and I believe MySQL may also crash and facing such problem?), so this is not a big issue.
```

Comment:
> In MySQL 8.0, this behavior is changed. The current maximum auto-increment counter value is written to the redo log each time it changes and saved to the data dictionary on each checkpoint. These changes make the current maximum auto-increment counter value persistent across server restarts.

https://dev.mysql.com/doc/refman/8.0/en/innodb-auto-increment-handling.html

And seems the actually allocated ID is larger than persisted max allocated ID? How do we avoid duplicated ID error after failover

Reply:
OK, mysql 8.0 does a good job!

> And seems the actually allocated ID is larger than persisted max allocated ID? 

Nope, the persisted max ID is larger than the to-be-allocated IDs @lance6716